### PR TITLE
TIP-1285: compute completeness and index products and ancestors

### PR DESCRIPTION
**Pull request by mmetayer**
group product subscribers:
- on save: compute completeness and index products and ancestors
- on delete: remove products from index and reindex ancestors

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        return [
+            StorageEvents::POST_SAVE => 'handleSingleProduct',
+            StorageEvents::POST_SAVE_ALL => 'handleMultipleProducts',
+        ];
+    }
+
+    public function handleSingleProduct(GenericEvent $event): void
+    {
+        if (false === ($event->getArguments()['unitary'] || true)) {
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: Think a notice is thrown if key unitary does not exist

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                    && get_class($product) !== 'Akeneo\Pim\WorkOrganization\Workflow\Component\Model\PublishedProduct';
+            }
+        );
+
+        $productIdentifiers = array_map(
+            function (ProductInterface $product): string {
+                return $product->getIdentifier();
+            },
+            $products
+        );
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: array_map and array_filter could be merged in one instruction/loop. ok it's micro optimization :smiley:

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php (hunk 1 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: That's why the `||` is here :wink:

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: Don't see what's the `|| true` is changing. The `?? true` is ok, but for me the `||` does not handle the case where the key `unitary` does not exist

Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: oops 🙄

Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: http://sandbox.onlinephpfunctions.com/code/bb2b02e8ac13084dd24f9cddf4763cdee53bf35f

Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: Yeah yeah, I'm ashamed enough, no need to push it... :smile:

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+use Akeneo\Pim\Enrichment\Component\Product\Model\ProductInterface;
+use Akeneo\Tool\Component\StorageUtils\StorageEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+/**
+ * This subscriber is responsible for
+ * - computing and persisting product completeness
+ * - reindexing products
+ * - reindexing ancestor product models
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]:
```suggestion
 * - reindexing ancestor product models as the completeness could change and impact the product model field 'all_complete' or 'all_incomplete' in the ES projection
```

File: src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php (hunk 2 above)
Comment (mmetayer) [reply src/Akeneo/Pim/Enrichment/Bundle/EventSubscriber/Product/OnSave/ComputeProductsAndAncestorsSubscriber.php]: would be less readable IMO

**Pull request merged**